server/src/main/java/org/elasticsearch/index/mapper/TimestampFieldMapper.java

```diff
@@ -0,0 +1,255 @@
+                new TimestampFieldType(),
+                path
+            );
+        }
+    }
+
+    public static class TypeParser implements MetadataFieldMapper.TypeParser {
+
+        @Override
+        public MetadataFieldMapper.Builder<?> parse(String name,
+                                                    Map<String, Object> node,
+                                                    ParserContext parserContext) throws MapperParsingException {
+            Builder builder = new Builder();
+            for (Iterator<Map.Entry<String, Object>> iterator = node.entrySet().iterator(); iterator.hasNext();) {
+                Map.Entry<String, Object> entry = iterator.next();
+                String fieldName = entry.getKey();
+                Object fieldNode = entry.getValue();
+                if (fieldName.equals("path")) {
+                    builder.setPath((String) fieldNode);
+                    iterator.remove();
+                }
+            }
+            return builder;
+        }
+
+        @Override
+        public MetadataFieldMapper getDefault(ParserContext parserContext) {
+            return new TimestampFieldMapper(Defaults.TIMESTAMP_FIELD_TYPE,
+                new TimestampFieldType(), null);
+        }
+    }
+
+    private final String path;
+
+    private TimestampFieldMapper(FieldType fieldType, MappedFieldType mappedFieldType, String path) {
+        super(fieldType, mappedFieldType);
+        this.path = path;
+    }
+
+    public void validate(DocumentFieldMappers lookup) {
+        if (path == null) {
+            // not configured, so skip the validation
+            return;
+        }
+
+        Mapper mapper = lookup.getMapper(path);
+        if (mapper == null) {
+            throw new IllegalArgumentException("timestamp meta field's field_name [" + path + "] points to a non existing field");
+        }
+
+        if (DateFieldMapper.CONTENT_TYPE.equals(mapper.typeName()) == false &&
+            DateFieldMapper.DATE_NANOS_CONTENT_TYPE.equals(mapper.typeName()) == false) {
+            throw new IllegalArgumentException("timestamp meta field's field_name [" + path + "] is of type [" +
+                mapper.typeName() + "], but [" + DateFieldMapper.CONTENT_TYPE + "," + DateFieldMapper.DATE_NANOS_CONTENT_TYPE +
+                "] is expected");
+        }
+
+        DateFieldMapper dateFieldMapper = (DateFieldMapper) mapper;
+        if (dateFieldMapper.fieldType().isSearchable() == false) {
+            throw new IllegalArgumentException("timestamp meta field's field_name [" + path + "] is not indexed");
+        }
+        if (dateFieldMapper.fieldType().hasDocValues() == false) {
+            throw new IllegalArgumentException("timestamp meta field's field_name [" + path + "] doesn't have doc values");
+        }
+
+        // Validate whether disallowed mapping attributes have been specified on the field this meta field refers to:
+        try (XContentBuilder builder = jsonBuilder()) {
```

Comment:
To me it'd be cleaner to avoid working with the raw JSON and just check specific attributes like 'null value' and 'ignore malformed'. We can add getters to `DateFieldMapper` if it makes it easier.

Reply:
This is a catch all validation, if there is any attribute specified but type, meta or format then fail.
I prefer to keep this logic in the case we ever forget to check an (future new) attribute.
The idea is that for least for now in data streams we want to lock down what can be configured on the
timestamp field.

However I can add specific checks for null value and ignore malformed for clearer error messages.
Like the check for whether the field is searchable and has doc values.